jobs/integration/validation.py

```diff
@@ -2306,68 +2372,61 @@ async def test_nagios(model, nagios):
     """
 
     log("starting nagios test")
-    masters = model.applications["kubernetes-control-plane"]
-
-    # 1) break all the things
-    log("breaking api server")
-    await masters.set_config({"api-extra-args": "broken=true"})
-
-    # 2) make sure nagios is complaining for kubernetes-control-plane
-    #    AND kubernetes-worker
-    log("Verifying complaints")
-    while True:
-        alerts = await nagios.wait_for_alerts(critical=lambda c: len(c) > 0)
-        criticals = alerts["critical"]
+    control_plane = model.applications["kubernetes-control-plane"]
 
-        if criticals:
-            found_master = []
-            found_worker = []
-            for c in criticals:
-                for link in c.find_all("a", recursive=False):
-                    if "kubernetes-control-plane" in link.string:
-                        found_master.append(link.string)
-                    elif "kubernetes-worker" in link.string:
-                        found_worker.append(link.string)
-            if found_master and found_worker:
-                log("Found critical errors:")
-                for s in found_master + found_worker:
-                    log(" - {}".format(s))
-                break
-        await asyncio.sleep(30)
+    try:
+        # 1) break all the things
+        log("breaking api server")
+        await control_plane.set_config({"api-extra-args": "broken=true"})
 
-    # 6) fix api and wait for settle
-    log("Fixing API server")
-    await masters.set_config({"api-extra-args": ""})
-    await nagios.wait_for_alerts(critical=lambda c: not c)
+        # 2) make sure nagios is complaining for kubernetes-control-plane
+        #    AND kubernetes-worker
+        log("Verifying complaints")
+        await retry_async_with_timeout(
+            nagios.critical_alerts_by_app,
+            (
+                "kubernetes-control-plane",
+                "kubernetes-worker",
+            ),
+            timeout_insec=10 * 60,
+            timeout_msg="Failed to find critical errors in control-plane and worker after 10m.",
+            retry_interval_insec=30,
+        )
```

Comment:
this gets run in a `try...finally`.  If the tests FAILS (due to a timeout maybe) we still want to restore the broken control-plane. 

next notice the test is effectively run in the `critical_alerts_by_app` method.  keep looking for critical alerts in the control plane and worker units or until the 10 minutes is up.  A timeout constitutes a test failure.  But don't leave the control plane broken after the failed test.